assignments/02-cli-constructive.md

```diff
@@ -55,63 +56,101 @@ $ touch rng.c
+form the build script for `foo`.
 
 ### Requirements
 
+This all might sound a little bit scary but we promise that explicitly laying
+out the behavior like this makes for easier implementation. You can think of it
+like a checklist.
+
 > Note: The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
 > "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this
 > document are to be interpreted as described in [RFC
 > 2119](https://datatracker.ietf.org/doc/html/rfc2119).
 
 You MUST:
 
+* Be able to build `foo` by running `./build.sh`
+* Only recompile a target if any of its dependencies' modified-time (mtime) is
+  later than the target's mtime
 * Write your program so that it runs under `sh` or `bash`
+* Write your program entirely in one file
 * Include a shebang line at the beginning of your script for either `sh` or
   `bash`
-* 
-
-Your output should be a list of files that are present in the directory but not
-in the snapshot, one per line. In other words, print out the files that have
-been added since the snapshot was taken. These should be valid paths relative
-to the given directory (so either `file1` or `./file1` is acceptable). You may
-find the `comm` utility useful in generating this list; see `man comm`.
-
-Please include a `#!` line at the beginning of your script that indicates it
-should be executed with either `/bin/sh` or `/bin/bash`. (`/bin/sh` points to
-different shells on different systems but is always guaranteed to be a
-POSIX-compliant shell; `/bin/bash` is always Bash specifically.) And use the
-error handling practices we encouraged in lecture.
-
-Your program may write data to files as part of generating its result. If it
-does, please use the `mktemp` utility to create them. `mktemp` generates a new
-file with a random name inside `/tmp`, a system directory designed to hold
-temporary files that don't need to stick around. We encourage you to remove any
-temporary files you use once your script finishes, but you will not lose points
-if you don't.
-
-## Using syscalls: write your own ls!
+* Use the error-handling practices we encouraged in lecture
+
+You MUST NOT:
+
+* Shell out to `make` or `watchman` or `entr` or any other software that would
+  do substantive work for you
+* Modify any of the C source files we provide to you
+
+You MAY:
+
+* Print the commands executed by your script as they are executed
+* Take an optional argument to `build.sh` to specify what target to build
+* Use (small) standard Unix utilities. When in doubt, ask on Piazza
+* Define functions to make your code more readable
+* Support modifying the build by setting `CC`, `CFLAGS`, `LDFLAGS`, and other
+  assorted environment variables
+* Write data to temporary files as part of your build script. If you do, you
+  MUST:
+  * Use the `mktemp` utility to create them
+  * Automatically clean up the temporary files when the build script finishes
```

Comment:
How strictly will you grade this in the rubric? If some command in their script fails unexpectedly and `set -e` bails, do the files still need to be deleted? What about if you Ctrl-C the script?

Basically asking if they need to know how to use `trap` to meet this requirement or not.